components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java

```diff
@@ -211,91 +199,76 @@ protected String getToken(String tokenEndPoint, String clientId, String clientSe
+        if (StringUtils.isBlank(token)) {
+            String errorMessage = "Access token is empty or null";
             if (logger.isDebugEnabled()) {
-                logger.debug("URL : " + tokenRequest.getLocationUri());
+                logger.debug(errorMessage);
             }
-            throw new ApplicationAuthenticatorException("MalformedURLException while sending access token request.", e);
-        } catch (
-                IOException e) {
-            throw new ApplicationAuthenticatorException("IOException while sending access token request.", e);
+            throw new ApplicationAuthenticatorException(errorMessage);
         }
         return token;
     }
 
     protected String getAuthorizationCode(HttpServletRequest request) throws ApplicationAuthenticatorException {
 
-        OAuthAuthzResponse authzResponse;
+        OAuthAuthzResponse authorizationResponse;
         try {
-            authzResponse = OAuthAuthzResponse.oauthCodeAuthzResponse(request);
-            return authzResponse.getCode();
+            authorizationResponse = OAuthAuthzResponse.oauthCodeAuthzResponse(request);
+            return authorizationResponse.getCode();
         } catch (OAuthProblemException e) {
             throw new ApplicationAuthenticatorException("Exception while reading authorization code.", e);
         }
     }
 
-    protected String sendRequest(String url) throws IOException {
-
-        BufferedReader bufferReader = null;
-        StringBuilder stringBuilder = new StringBuilder();
+    private OAuthClientResponse getOauthResponse(OAuthClient oAuthClient, OAuthClientRequest accessRequest)
+            throws ApplicationAuthenticatorException {
 
+        OAuthClientResponse oAuthResponse;
         try {
-            URLConnection urlConnection = new URL(url).openConnection();
-            bufferReader = new BufferedReader(
-                    new InputStreamReader(urlConnection.getInputStream(), StandardCharsets.UTF_8));
-
-            String inputLine = bufferReader.readLine();
-            while (inputLine != null) {
-                stringBuilder.append(inputLine).append(Oauth2GenericAuthenticatorConstants.NEW_LINE);
-                inputLine = bufferReader.readLine();
+            oAuthResponse = oAuthClient.accessToken(accessRequest);
+            return oAuthResponse;
+        } catch (OAuthSystemException | OAuthProblemException e) {
+            if (logger.isDebugEnabled()) {
+                logger.debug(e.getMessage());
             }
-        } finally {
-            IdentityIOStreamUtils.closeReader(bufferReader);
+            throw new ApplicationAuthenticatorException(e.getMessage());
         }
-        return stringBuilder.toString();
     }
 
     protected OAuthClientRequest buildTokenRequest(String tokenEndPoint, String clientId, String clientSecret,
-                                                  String state, String code, String redirectUri,
```

Comment:
Does this method extended from the AbstractAuthenticator class? Why we removed state parameter?

Reply:
No it's not.